internal/jobcontainers/jobcontainer.go

```diff
@@ -0,0 +1,409 @@
+package jobcontainers
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/hcs"
+	"github.com/Microsoft/hcsshim/internal/jobobject"
+	"github.com/Microsoft/hcsshim/internal/layers"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/queue"
+	"github.com/Microsoft/hcsshim/internal/schema1"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/windows"
+)
+
+type initProc struct {
+	initDoOnce sync.Once
+	proc       *JobProcess
+	initBlock  chan struct{}
+}
+
+// JobContainer represents a lightweight container composed from a job object.
+type JobContainer struct {
+	id             string
+	spec           *specs.Spec          // OCI spec used to create the container
+	job            *jobobject.JobObject // Object representing the job object the container owns
+	sandboxMount   string               // Path to where the sandbox is mounted on the host
+	m              sync.Mutex
+	closedWaitOnce sync.Once
+	init           initProc
+	startTimestamp time.Time
+	exited         chan struct{}
+	waitBlock      chan struct{}
+	waitError      error
+}
+
+var _ cow.ProcessHost = &JobContainer{}
+var _ cow.Container = &JobContainer{}
+
+func newJobContainer(id string, s *specs.Spec) *JobContainer {
+	return &JobContainer{
+		id:        id,
+		spec:      s,
+		waitBlock: make(chan struct{}),
+		exited:    make(chan struct{}),
+		init:      initProc{initBlock: make(chan struct{})},
+	}
+}
+
+// Create creates a new JobContainer from `s`.
+func Create(ctx context.Context, id string, s *specs.Spec) (_ cow.Container, err error) {
+	log.G(ctx).WithField("id", id).Debug("Creating job container")
+
+	if s == nil {
+		return nil, errors.New("Spec must be supplied")
+	}
+
+	if id == "" {
+		g, err := guid.NewV4()
+		if err != nil {
+			return nil, err
+		}
+		id = g.String()
+	}
+
+	if err := mountLayers(ctx, s); err != nil {
+		return nil, errors.Wrap(err, "failed to mount container layers")
+	}
+
+	volumeGUIDRegex := `^\\\\\?\\(Volume)\{{0,1}[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}(\}){0,1}\}(|\\)$`
+	if matched, err := regexp.MatchString(volumeGUIDRegex, s.Root.Path); !matched || err != nil {
+		return nil, fmt.Errorf(`invalid container spec - Root.Path '%s' must be a volume GUID path in the format '\\?\Volume{GUID}\'`, s.Root.Path)
+	}
+	if s.Root.Path[len(s.Root.Path)-1] != '\\' {
+		s.Root.Path += `\` // Be nice to clients and make sure well-formed for back-compat
+	}
+
+	container := newJobContainer(id, s)
+
+	// Create the job object all processes will run in.
+	options := &jobobject.Options{
+		Name:          id,
```

Comment:
Should we prefix this with something like "JobContainer_"?

Reply:
Yea might as well